ToolBox/LeanDataWriter.cs

```diff
@@ -104,28 +127,105 @@ public void Write(IEnumerable<BaseData> source)
             // Write the last file
             if (sb.Length > 0)
             {
-                var outputFile = ZipOutputFile(baseDirectory, lastTime);
+                var outputFile = GetZipOutputFileName(baseDirectory, lastTime);
                 WriteFile(outputFile, sb.ToString(), lastTime);
             }
         }
 
         /// <summary>
+        /// Write out the data in LEAN format (daily or hour resolutions)
+        /// </summary>
+        /// <param name="source">IEnumerable source of the data: sorted from oldest to newest.</param>
+        /// <remarks>This function performs a merge (insert/append/overwrite) with the existing Lean zip file</remarks>
+        private void WriteDailyOrHour(IEnumerable<BaseData> source)
+        {
+            var sb = new StringBuilder();
+            var lastTime = new DateTime();
+
+            // Determine file path
+            var baseDirectory = Path.Combine(_dataDirectory, _securityType.ToString().ToLower(), _market);
+
+            var outputFile = GetZipOutputFileName(baseDirectory, lastTime);
+
+            // Load new data rows into a SortedDictionary for easy merge/update
+            var newRows = new SortedDictionary<DateTime, string>(source.ToDictionary(x => x.Time, GenerateFileLine));
+            SortedDictionary<DateTime, string> rows;
+
+            if (File.Exists(outputFile))
+            {
+                // If file exists, we load existing data and perform merge
+                rows = LoadHourlyOrDailyFile(outputFile);
+                foreach (var kvp in newRows)
+                {
+                    rows[kvp.Key] = kvp.Value;
```

Comment:
I think we want to load the hourly/daily file first, and then overly the newly created data. My thought is this, if the date in the old file is corrupt then it will never get replaced.